credentials/credentials.go

```diff
@@ -71,15 +73,10 @@ type Credentials interface {
 // TransportAuthenticator defines the common interface all supported transport
 // authentication protocols (e.g., TLS, SSL) must implement.
 type TransportAuthenticator interface {
-	// Dial connects to the given network address using net.Dial and then
+	// Dial connects to the given network address using dialer and then
 	// does the authentication handshake specified by the corresponding
 	// authentication protocol.
-	Dial(network, addr string) (net.Conn, error)
-	// DialWithDialer connects to the given network address using
-	// dialer.Dial does the authentication handshake specified by the
-	// corresponding authentication protocol. Any timeout or deadline
-	// given in the dialer apply to connection and handshake as a whole.
-	DialWithDialer(dialer *net.Dialer, network, addr string) (net.Conn, error)
+	Dial(dialer func(string, time.Duration) (net.Conn, error), addr string, timeout time.Duration) (net.Conn, error)
```

Comment:
This signature is weird. We're passing in a function and its two arguments. Maybe instead of Dial this should be `Handshake(net.Conn) error` instead, and then the caller can do the dialing?